crates/sui-core/src/consensus_manager/mod.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use crate::consensus_manager::mysticeti_manager::MysticetiManager;
+use crate::consensus_manager::narwhal_manager::{
+    NarwhalConfiguration, NarwhalManager, NarwhalManagerMetrics,
+};
+use fastcrypto::traits::KeyPair;
+use mysten_metrics::RegistryService;
+use narwhal_executor::ExecutionState;
+use narwhal_worker::TransactionValidator;
+use std::path::PathBuf;
+use std::sync::Arc;
+use sui_config::{ConsensusConfig, NodeConfig};
+
+pub mod mysticeti_manager;
+pub mod narwhal_manager;
+
+/// An enum to easily differentiate between the chosen consensus engine
+pub enum Manager {
+    Narwhal(NarwhalManager),
+    Mysticeti(MysticetiManager),
+}
+
+impl Manager {
+    /// Create a new narwhal manager and wrap it around the Manager enum
+    pub fn narwhal(
+        config: &NodeConfig,
+        consensus_config: &ConsensusConfig,
+        registry_service: &RegistryService,
+    ) -> Self {
+        let narwhal_config = NarwhalConfiguration {
+            primary_keypair: config.protocol_key_pair().copy(),
+            network_keypair: config.network_key_pair().copy(),
+            worker_ids_and_keypairs: vec![(0, config.worker_key_pair().copy())],
+            storage_base_path: consensus_config.db_path().to_path_buf(),
+            parameters: consensus_config.narwhal_config().to_owned(),
+            registry_service: registry_service.clone(),
+        };
+
+        let metrics = NarwhalManagerMetrics::new(&registry_service.default_registry());
+
+        Manager::Narwhal(NarwhalManager::new(narwhal_config, metrics))
+    }
+
+    // Starts the underneath consensus manager by the given inputs
+    pub async fn start<State, StateInitializer, TxValidator: TransactionValidator>(
```

Comment:
I think we don't need parametrize this, since it's sui specific code.

So, for example for `TxValidation`, it should just take instance of `SuiTxValidator`. Then we can implement `mysticeti::BlockValidator` that @longbowlu wrote for the `SuiTxValidator` and this way we will be able to pass it to mysticeti.

Same for ConsensusHandler - we can just pass concrete instance here.

Reply:
We can absolutely do that - just kept it generic for now as it wasn't clear to me how things would evolve. For example maybe for Mysticeti we would craft a separate ConsensusHandler - not that likely but still. Changing to concrete implementation.